src/Neo/SmartContract/Native/NeoToken.cs

```diff
@@ -327,13 +327,34 @@ public BigInteger UnclaimedGas(DataCache snapshot, UInt160 account, uint end)
             return CalculateBonus(snapshot, state, end);
         }
 
+        [ContractMethod(Hardfork.HF_Echidna, RequiredCallFlags = CallFlags.States | CallFlags.AllowCall | CallFlags.AllowNotify)]
+        private async ContractTask OnNEP17Payment(ApplicationEngine engine, UInt160 from, BigInteger amount, StackItem data)
+        {
+            if (engine.CallingScriptHash != GAS.Hash)
+                throw new InvalidOperationException("only GAS is accepted");
+
+            if ((long)amount != GetRegisterPrice(engine.SnapshotCache))
+                throw new ArgumentException("incorrect GAS amount for registration");
+
+            var pubkey = ECPoint.DecodePoint(data.GetSpan(), ECCurve.Secp256r1);
+
+            await GAS.Burn(engine, Hash, amount);
+
+            RegisterInternal(engine, pubkey);
```

Comment:
I raised this question in https://github.com/nspcc-dev/neo-go/pull/3700#discussion_r1856353189, but let's collect some more opinions on this topic. Please, read the link and share what do you think about the order of these method calls.

Reply:
@shargon

Reply:
Fixed. Let's burn after the state change. Also fixed registration handling (and added a test for it).